File: findpowder.m
function f = findpowder(x,img_exp,plotflag)
% to load data
% ccd = double(h5read('h5/scan_5_028866.h5','/entry/data/data',[1,1,1],[1028,1062,1]));
% ccd = ccd.*(double(ccd<4e9));
% ccd = ccd.*(double(ccd>0));
% ccd = ccd';
% for ii = 2:20;ccd1 = double(h5read('h5/scan_5_028866.h5','/entry/data/data',[1,1,ii],[1028,1062,1]));ccd1 = ccd1.*(double(ccd1<4e9));ccd1 = ccd1.*(double(ccd1>0));ccd = ccd1'+ccd;end
% to view and fit
% findpowder(1e5*[1.8 -0.375 -0.30 .00037],ccd,1);
% x=fminsearch(@(x) findpowder(x,ccd,0),1e5*[1.8 -0.375 -0.30 .00037]);
% findpowder(x,ccd,1);
% x = [1.806360774795135e+05,-3.729935571548389e+04,-2.859329871988752e+04,36.970724978683790]

f1 = qbin_pilatus(img_exp, x(1), x(2), x(3), x(4), x(5), 0, 0, 0, plotflag);


if(plotflag)
    figure(701);imagesc(f1.powder);title('Thresholded experiment');axis image
    figure(702);imagesc(f1.theory);title('Theory');axis image
end
% two theta, R, (gam), X, Y
% input - R X Y 2th
if(plotflag)
    f=f1;
else
    f = f1.error;
end
end

File: findpowder.py
"""
Nov 11, 2024 Developed by Rui Liu @ANL
rui.liu@anl.gov
This module is for detector callibration at 26ID at APS
"""
import numpy as np
import matplotlib.pyplot as plt
import imageio.v2 as imageio
import os
from qbin_pilatus import qbin_pilatus

# Function to save results
def save_results(dataout, output_folder="Calibration Results"):
    """
    Save the results to CSV files in the specified output folder.

    Parameters:
    - dataout: Dictionary containing the results (e.g., gamma, two_theta, etc.)
    - output_folder: Folder where results will be saved
    """
    # Create the folder if it doesn't exist
    os.makedirs(output_folder, exist_ok=True)

    # Save each key in the dictionary to a CSV file
    np.savetxt(os.path.join(output_folder, "gamma.csv"), dataout["gamma"], delimiter=",")
    np.savetxt(os.path.join(output_folder, "gamma_mot.csv"), dataout["gamma"], delimiter=",")
    np.savetxt(os.path.join(output_folder, "twotheta.csv"), dataout["two_theta"], delimiter=",")
    np.savetxt(os.path.join(output_folder, "twotheta_mot.csv"), dataout["two_theta"], delimiter=",")

    print(f"Results saved in folder: {output_folder}")

def findpowder(x, img_exp, plotflag=True):
    """
    Function to find powder diffraction calibration.

    Parameters:
    - x: List of parameters [R, Tthdet, Gamdet, Xdet, Ydet]
    - img_exp: Experimental image (2D array)
    - plotflag: Whether to plot the results

    Returns:
    - f: Dictionary containing results if plotflag=True, otherwise the error value
    """
    # Unpack the parameters from x
    Tthdet, R, Gamdet, Xdet, Ydet = x

    # Call qbin_pilatus with the unpacked parameters
    f1 = qbin_pilatus(img_exp, Tthdet, R, Gamdet, Xdet, Ydet, 0, 0, 0, plotflag)

    # If plotflag is True, display the images
    if plotflag:
         # Plot the thresholded experimental data
        plt.figure(701)
        plt.imshow(f1["powder"], cmap="gray")
        plt.title("Thresholded Experiment")
        plt.axis("image")
        plt.colorbar()
        plt.show()

        # Plot the theoretical data
        plt.figure(702)
        plt.imshow(f1["theory"], cmap="gray")
        plt.title("Theory")
        plt.axis("image")
        plt.colorbar()
        plt.show()

        return f1  # Return the full result dictionary

    # If plotflag is False, return only the error
    return f1["error"]

# Example usage
if __name__ == "__main__":
    # Define the image path
    image_path = 'Images/CAP_1.tiff'

    # Load the experimental image
    img_exp = imageio.imread(image_path)

    # Example parameters [R, Tthdet, Gamdet, Xdet, Ydet]
    x = [20, 150000, 17, -18000, 2500]

    # Call the findpowder function with plotting enabled
    result = findpowder(x, img_exp, plotflag=True)
    # Save the results for comparison
    save_results(result)

    # Output the result for debugging
    # print(result)